psi4/driver/procrouting/scf_proc/efp.py

```diff
@@ -0,0 +1,187 @@
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License along
+# with Psi4; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# @END LICENSE
+#
+
+from __future__ import division
+
+import numpy as np
+
+from psi4 import core
+
+
+def get_qm_atoms_opts(mol):
+    """Provides list of coordinates of quantum mechanical atoms from
+    psi4.core.Molecule `mol` to pylibefp.core.efp() `efpobj`. Also
+    converts from `read_options("EFP"` to pylibefp opts dictionary.
+
+    """
+    efpobj = mol.EFP
+
+    ptc = []
+    coords = []
+    for iat in range(mol.natom()):
+        ptc.append(mol.charge(iat))
+        coords.append(mol.x(iat))
+        coords.append(mol.y(iat))
+        coords.append(mol.z(iat))
+
+    # set options
+    # * 'chtr', 'qm_exch', 'qm_disp', 'qm_chtr' may be enabled in a future libefp release
+    opts = {}
+    for opt in ['elst', 'exch', 'ind', 'disp',
+                'elst_damping', 'ind_damping', 'disp_damping']:
+        psiopt = 'EFP_' + opt.upper()
+        if core.has_option_changed('EFP', psiopt):
+            opts[opt] = core.get_option('EFP', psiopt)
+    for opt in ['elst', 'ind']:
+        psiopt = 'EFP_QM_' + opt.upper()
+        if core.has_option_changed('EFP', psiopt):
+            opts['qm_' + opt] = core.get_option('EFP', psiopt)
+
+    return ptc, coords, opts
+
+
+def modify_Fock_permanent(mol, mints, verbose=1):
+    """Computes array of the EFP contribution to the potential felt by
+    QM atoms due to permanent EFP moments. Used for SCF procedure.
+
+    Parameters
+    ----------
+    mol : :py:class:`psi4.core.Molecule`
+        Source of quantum mechanical atoms. As its `EFP` member data, contains
+        a :py:class:`pylibefp.core.efp` object that is the source and computer
+        of EFP fragments.
+    mints : `psi4.core.MintsHelper`
+        Integral computer.
+    verbose : int, optional
+        Whether to print out multipole coordinates and values. 0: no printing.
+        1: print charges and dipoles. 2: additionally print quadrupoles and octupoles.
+
+    Returns
+    -------
+    ndarray
+        (nbf, nbf) EFP charge through octupole contribution to the potential
+
+    """
+    # get composition counts from libefp
+    efpobj = mol.EFP
+    nfr = efpobj.get_frag_count()
+    natoms = efpobj.get_frag_atom_count()
+
+    # get multipoles count, pos'n, values from libefp
+    #   charge + dipoles + quadrupoles + octupoles = 20
+    nmp = efpobj.get_multipole_count()
+    xyz_mp = np.asarray(efpobj.get_multipole_coordinates(verbose=verbose)).reshape(nmp, 3)
+    val_mp = np.asarray(efpobj.get_multipole_values(verbose=verbose)).reshape(nmp, 20)
+
+    #                    0  X  Y  Z  XX   YY   ZZ   XY   XZ   YZ
+    prefacs = np.array([ 1, 1, 1, 1, 1/3, 1/3, 1/3, 2/3, 2/3, 2/3,
+        1/15, 1/15, 1/15, 3/15, 3/15, 3/15, 3/15, 3/15, 3/15, 6/15])
```

Comment:
I know we're dropping py2 support, but I don't think this division works correctly there.  If we're dropping it now, that's totally find with me, else we should probably have 1.0/3, etc.

Reply:
it's got `from __future__ import division` so I think it's ok.

Reply:
Oh, good call.  As you were!